File: matlab/autd3/Modulation.m
%{
%File: Modulation.m
%Project: autd3-matlab
%Created Date: 07/06/2022
%Author: Shun Suzuki
%-----
%Last Modified: 11/06/2022
%Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
%-----
%Copyright (c) 2022 Shun Suzuki. All rights reserved.
%
%}

classdef Modulation < Header

    properties
        sampling_frequency_division
    end

    methods

        function obj = Modulation()
            obj = obj@Header();
        end

        function set.sampling_frequency_division(obj, value)
            obj.sampling_frequency_division = value;
            calllib('autd3capi', 'AUTDModulationSetSamplingFrequencyDivision', obj.ptr, value);
        end

        function div = get.sampling_frequency_division(obj)
            div = calllib('autd3capi', 'AUTDModulationSamplingFrequencyDivision', obj.ptr);
        end

        function freq = sampling_frequency(obj)
            freq = calllib('autd3capi', 'AUTDModulationSamplingFrequency', obj.ptr);
        end

        function delete(obj)

            if obj.ptr.Value ~= 0
                calllib('autd3capi', 'AUTDDeleteModulation', obj.ptr);
                obj.ptr = libpointer('voidPtr', 0);
            end

        end

    end

end

File: nim/src/autd3/modulation.nim
# File: modulation.nim
# Project: autd3
# Created Date: 13/06/2022
# Author: Shun Suzuki
# -----
# Last Modified: 13/06/2022
# Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
# -----
# Copyright (c) 2022 Shun Suzuki. All rights reserved.
#

import header
import native_methods/autd3capi

type Modulation* = object of Header

proc `=destroy`(modulation: var Modulation) =
    if (modulation.p != nil):
        AUTDDeleteModulation(modulation.p)
        modulation.p = pointer(nil)

proc samplingFrequencyDivision*(modulation: Modulation): uint32 =
    AUTDModulationSamplingFrequencyDivision(modulation.p)

proc `samplingFrequencyDivision=`*(modulation: Modulation, value: uint32) =
    AUTDModulationSetSamplingFrequencyDivision(modulation.p, value)

proc samplingFrequency*(modulation: Modulation): float64 =
    AUTDModulationSamplingFrequency(modulation.p)
